### ISSUE TITLE
False negative with dict literal on strict mode

### ISSUE THREAD

**Issue by rayansostenes** [microsoft/pyright#6514]
```python
def test(x: dict[str, int]):
    ...

d1 = {}
test(d1)

d2 = dict()
test(d2)
```

Error is correctly identified when using `dict`, but not when the dict is constructed using the literal syntax.

```
Type of "d2" is partially unknown
  Type of "d2" is "dict[Unknown, Unknown]"  (reportUnknownVariableType)
Argument type is partially unknown
  Argument corresponds to parameter "x" in function "test"
  Argument type is "dict[Unknown, Unknown]"  (reportUnknownArgumentType)
```

[Playground Link](https://pyright-play.net/?strict=true&locale=en&code=FAEwpgZgBALmDOMAUAPAXFEBLAxjA2ogE4A0UWAdjALoCUawUTUAdG8KAIxQC8UA3gF9gcREhCdaHEACZemXMimjks2kA)

**Comment by erictraut** [comment]
This behavior is intended. When you construct a `dict` using an empty dict expression, pyright internally treats this as different from a `dict` constructed in other ways. This special-casing allows for improved type inference in common situations.

```python
def func(cond: bool):
    if cond:
        d1 = {}
    else:
        d1 = {"a": 1}
    reveal_type(d1) # dict[str, int]
```

When you use the explicit `dict` constructor call, the type comes from the typeshed definitions, and pyright doesn't generally override or special-case typeshed definitions because this causes many problems.

I recommend using dict expressions rather than explicit calls to a `dict` constructor where possible. The same goes for `list`, `set` and `tuple` expressions.

**Comment by rayansostenes** [comment]
@erictraut 
Still feels like a bug to be, what should I do to get an error on a case like this:

```python
from typing import Iterable, reveal_type


def test(x: dict[str, int]) -> Iterable[tuple[str, int]]:
    for s, i in x.items():
        yield s.capitalize(), i + 2

d1 = {}
for i, n in enumerate('ABC'):
    reveal_type(i)
    reveal_type(n)
    d1[i] = n

reveal_type(d1)
list(test(d1))
```

pyright Output:
```
Type of "i" is "int"
Type of "n" is "str"
Type of "d1" is "dict[Unknown, Unknown]"
```

Runtime result:
```
Traceback (most recent call last):
  File "<stdin>", line 15, in <module>
  File "<stdin>", line 6, in test
AttributeError: 'int' object has no attribute 'capitalize'
```

**Comment by erictraut** [comment]
I don't see a way to generate an error here without creating many false positives. It's very common to pass `{}` or `[]` as an argument to a function. This shouldn't generate an error even in strict mode. When we're forced to make a tradeoff between false positives and false negatives, we generally try to avoid false positives. This particular case involves a strict-mode check, so the calculus changes a bit, but I'm still concerned about the noise this would create in code bases that are using strict-mode type checking.

**Comment by erictraut** [comment]
I'm going to reopen this issue and give it more thought. I have some ideas that I want to explore that may allow for more precisely distinguishing between the positive and negative cases.

**Comment by erictraut** [comment]
I've found a way to eliminate many of the false positives. This change will be in the next release. I'm still somewhat concerned about the noise this will cause in "strict mode" type bases, but the fix is straightforward in most cases, and it will be a one-time fix.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.338, which I just published. It will also be included in a future release of pylance.
